examples/sleuth.py

```diff
@@ -0,0 +1,299 @@
+            f'The following question is about this report:\n' +
+            PRIMARY_ASSET + '\n'
+        )
+        question = question_context + question
+        for message in agent(question):
+            if message.role=='user':
+                status = 'User agent asked a question or provided feedback to the LLM.  Awaiting LLM response...'
+                update = [format_message(message), None]
+            elif message.role=='function':
+                status = 'A query was run against the database.  Awaiting LLM response...'
+                update = [format_message(message), None]
+            elif message.role=='assistant':
+                if 'Final Answer:' in message.content:
+                    status = 'The final answer has been determined.'
+                else:
+                    status = 'The assistant responded.  Awaiting LLM next response...'
+                update = [None, format_message(message)]
+            chat_history.append(update)
+            yield status, chat_history
+
+        for db in DATABASES:
+            DATABASES[db]['connection'].close()
+            DATABASES[db]['connection'] = None
+
+    def answer_wrapper(chat_history):
+        final_message = chat_history[-1][1]
+        if final_message is None:
+            response = 'It appears the agent couldn\'t answer the questions.'
+        elif 'Final Answer:' in final_message:
+            response = final_message.split('Final Answer:')[1].strip()
+            response = response.split('_<name')[0].strip()
+        else:
+            response = 'It appears the agent couldn\'t answer the questions.'
+
+        return response
+
+    with gr.Blocks() as demo:
+        gr.Markdown('# Chat About Techtonic Tech, Inc\'s P&L')
+        gr.Markdown('```\n' + PRIMARY_ASSET + '```')
+
+        question = gr.Textbox(
+            value=DEFAULT_QUESTION, label='Ask a question')
+        with gr.Row():
+            with gr.Column():
+                ask_button = gr.Button('Ask')
+            with gr.Column():
+                clear_button = gr.ClearButton()
+        status = gr.Textbox(
+            value='Status will appear here', label='Agent status', interactive=False
+        )
+        final_answer = gr.Textbox(value='', label='Answer', interactive=False)
+
+        chatbot = gr.Chatbot(label='Chatbot message log')
+
+        gr.Markdown(
+            'You have access to four databases:\n'
+            '\n'
+            '1. Expense database with high-level expenses per category.\n'
+            '2. Employee compensation database with the cost of employee taxes, salaries and bonuses.\n'
+            '3. Competition database with competitor\'s rough financial and market data for the year.\n'
+            '4. Customer feedback database with customer feedback by category.'
+        )
+
+        clear_button.add([question, chatbot])
+
+        question.submit(
+            new_question_wrapper,
+            outputs=final_answer) \
+        .then(
```

Comment:
You can avoid the backslashes and PEP warnings about missing indentation by moving the close paren to the new line with the .then:

```
        question.submit(
            new_question_wrapper,
            outputs=final_answer
        ).then(
            agent_wrapper,
            [question, status, chatbot],
            [status, chatbot]
        ).then(
            answer_wrapper, chatbot, final_answer)

```